lib/apiservers/engine/backends/volume.go

```diff
@@ -96,12 +105,103 @@ func (v *Volume) Volumes(filter string) ([]*types.Volume, []string, error) {
 		}
 		volume := NewVolumeModel(vol, volumeMetadata.Labels)
 		volumes = append(volumes, volume)
 	}
-	return volumes, nil, nil
+
+	// Parse and validate filter args
+	volumeFilters, err := filters.FromParam(filter)
+	if err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+	if err := volumeFilters.Validate(acceptedVolumeFilters); err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	if volumeFilters.Len() == 0 {
+		return volumes, nil, nil
+	}
+
+	// If the dangling filter is specified, gather required items beforehand
+	dangling := false
+	joinedVolumes := make(map[string]struct{})
+	if volumeFilters.Include("dangling") {
+
+		// Validate dangling filter's usage (per Docker code)
+		if volumeFilters.ExactMatch("dangling", "true") || volumeFilters.ExactMatch("dangling", "1") {
+			dangling = true
+		} else if !volumeFilters.ExactMatch("dangling", "false") && !volumeFilters.ExactMatch("dangling", "0") {
+			return nil, nil, fmt.Errorf("invalid filter 'dangling=%s'", volumeFilters.Get("dangling"))
+		}
+
+		// Get all containers and use containers' metadata to obtain non-dangling (joined) volumes
+		conts, err := getAllContainers()
+		if err != nil {
+			return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+		}
+		for i := range conts {
+			for _, vol := range conts[i].VolumeConfig {
+				joinedVolumes[vol.Name] = struct{}{}
+			}
+		}
+	}
+
+	outputVols := filterVolumes(volumes, joinedVolumes, volumeFilters, dangling)
+	return outputVols, nil, nil
+}
+
+// filterVolumes returns volumes that meet the filtering criteria.
+func filterVolumes(volumes []*types.Volume, joinedVolumes map[string]struct{}, volumeFilters filters.Args, dangling bool) []*types.Volume {
+	var outputVols []*types.Volume
+
+	for _, vol := range volumes {
+
+		if volumeFilters.Include("dangling") {
+			_, joined := joinedVolumes[vol.Name]
+			if (!dangling || joined) && (dangling || !joined) {
+				continue
+			}
+		}
+
+		if volumeFilters.Include("name") {
+			if !volumeFilters.Match("name", vol.Name) {
+				continue
+			}
+		}
+
+		if volumeFilters.Include("driver") {
+			if !volumeFilters.ExactMatch("driver", vol.Driver) {
+				continue
+			}
+		}
+
+		outputVols = append(outputVols, vol)
+	}
+
+	return outputVols
+}
+
+// getAllContainers obtains all containers from the portlayer, akin to `docker ps -a`.
+func getAllContainers() ([]*models.ContainerInfo, error) {
+	client := PortLayerClient()
+	if client == nil {
+		return nil, fmt.Errorf("failed to get a portlayer client")
+	}
+
+	all := true
+	cons, err := client.Containers.GetContainerList(containers.NewGetContainerListParamsWithContext(ctx).WithAll(&all))
+	if err != nil {
+		switch err := err.(type) {
+		case *containers.GetContainerListInternalServerError:
+			return nil, fmt.Errorf("error invoking GetContainerList: %s", err.Payload.Message)
```

Comment:
Not sure why we're doing this.  GetCOntainerListInt... is already an error.  Do we need to make it a generic error?